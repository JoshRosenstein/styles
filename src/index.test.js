import styles from './index'

afterEach(() => {
  document.getElementsByTagName('html')[0].innerHTML = ''
})

describe('Usage', () => {
  it('returns single class name', () => {
    const style = styles.configure()({
      flex: {display: 'flex'},
      block: {display: 'block'},
      inline: 'display: inline;',
    })

    expect(style('flex')).toMatchSnapshot()
    expect(style('flex', 'block', 'inline')).toMatchSnapshot()
    expect(style({flex: true, block: false, inline: true})).toMatchSnapshot()
  })

  it('returns empty string when falsy', () => {
    const style = styles.configure()({
      flex: {display: 'flex'},
    })

    let name = style(false)
    expect(typeof name).toBe('string')
    expect(name.length).toBe(0)

    name = style(false, null, undefined, 0, {flex: false})
    expect(typeof name).toBe('string')
    expect(name.length).toBe(0)
  })

  it('ignores unknown keys', () => {
    const style = styles.configure()({
      flex: {display: 'flex'},
    })

    let name = style('noop')
    expect(typeof name).toBe('string')
    expect(name.length).toBe(0)

    name = style({noop: true})
    expect(typeof name).toBe('string')
    expect(name.length).toBe(0)
  })

  it('allows unitless object values', () => {
    const style = styles.configure()({
      box: {width: 200, height: '200px'},
    })

    style('box')

    for (let element of document.querySelectorAll(`style[data-dash]`)) {
      expect(element).toMatchSnapshot('200x200')
    }
  })

  it('adds styles by order of definition when called', () => {
    const style = styles.configure({prefix: false})({
      inline: 'display: inline;',
      flex: {display: 'flex'},
      block: {display: 'block'},
    })

    style('flex', 'block', 'inline')

    for (let element of document.querySelectorAll(`style[data-dash]`)) {
      expect(element).toMatchSnapshot('flex, block, inline')
    }

    styles.cache.clear()
    styles.sheet.flush()
    style({flex: true, block: true, inline: true})

    for (let element of document.querySelectorAll(`style[data-dash]`)) {
      expect(element).toMatchSnapshot('flex, block, inline')
    }
  })

  it('supports internally nested styles w/ functions', () => {
    const style = styles.configure({prefix: false})({
      blue: style => `
        color: blue;
        ${style('red')} {
          color: purple;
        }
      `,
      red: 'color: red;',
    })

    style('blue')

    for (let element of document.querySelectorAll(`style[data-dash]`)) {
      expect(element).toMatchSnapshot(style('red'))
    }
  })

  it('allows multiple arguments in styles()', () => {
    const style = styles.configure({prefix: false})(
      {
        blue: style => `
        color: blue;
        ${style('red')} {
          color: purple;
        }
      `,
      },
      {
        red: 'color: red;',
      }
    )

    expect(style('blue')).toMatchSnapshot()
    expect(style('red')).toMatchSnapshot()
  })

  it('allows comments', () => {
    const style = styles.configure()({
      flex: `
        /* this is a flex style */
        display: flex;
      `,
    })

    expect(style('flex')).toMatchSnapshot()
  })

  it('adds dev labels', () => {
    let prevEnv = process.env.NODE_ENV
    process.env.NODE_ENV = 'development'
    const style = styles.configure()({
      flex: `display: flex;`,
      block: `display: block`,
      inline: `display: inline;`,
    })

    expect(style('flex')).toMatchSnapshot('-flex')
    expect(style('flex', 'inline')).toMatchSnapshot('-flex-inline')
    expect(style('flex', {inline: false, block: true})).toMatchSnapshot(
      '-flex-block'
    )
    process.env.NODE_ENV = prevEnv
  })
})

describe('Exceptions', () => {
  it('throws error for extract methods', () => {
    const style = styles.configure()({
      flex: {display: 'flex'},
    })

    expect(() => {
      style.extract()
    }).toThrowErrorMatchingSnapshot()

    expect(() => {
      style.extractTags()
    }).toThrowErrorMatchingSnapshot()
  })

  it('throws for unterminated comments', () => {
    const style = styles.configure()({
      flex: `
        /* this is a flex style with an unterminated comment ;)
        display: flex;
      `,
    })

    expect(() => {
      style('flex')
    }).toThrowErrorMatchingSnapshot()
  })
})
