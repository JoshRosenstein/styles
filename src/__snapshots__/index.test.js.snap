// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exceptions throws for unterminated comments 1`] = `"Your styles have an unterminated comment (\\"/*\\" without corresponding \\"*/\\")."`;

exports[`styles() adds dev labels: -flex 1`] = `"-ui-1ut9bc3-flex"`;

exports[`styles() adds dev labels: -flex-block 1`] = `"-ui-14t38od-flex-block"`;

exports[`styles() adds dev labels: -flex-inline 1`] = `"-ui-broszf-flex-inline"`;

exports[`styles() adds styles by order of definition when called: flex, block, inline 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-qaw0al{display:flex;display:block;display:inline;}
</style>
`;

exports[`styles() adds styles by order of definition when called: flex, block, inline 2`] = `
<style
  data-dash="-ui"
>
  
  .-ui-qaw0al{display:flex;display:block;display:inline;}
</style>
`;

exports[`styles() allows comments 1`] = `"-ui-xmqurs"`;

exports[`styles() allows default styles 1`] = `
NodeList [
  <style
    data-dash="-ui"
  >
    
    .-ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  </style>,
]
`;

exports[`styles() allows multiple arguments 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-qaw0al{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;display:block;display:inline;}
</style>
`;

exports[`styles() allows style functions in arguments 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-qaw0al{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;display:block;display:inline;}
</style>
`;

exports[`styles() allows unitless object values: 200x200 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-haiabg{width:200px;height:200px;}
</style>
`;

exports[`styles() has a default style that is always applied first 1`] = `
NodeList [
  <style
    data-dash="-ui"
  >
    
    .-ui-14t38od{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;display:block;}
  </style>,
]
`;

exports[`styles() only applies default style once 1`] = `
NodeList [
  <style
    data-dash="-ui"
  >
    
    .-ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  </style>,
]
`;

exports[`styles() passes variables to style callbacks 1`] = `
Object {
  "colors": Object {
    "bg": "var(--ui-colors-bg)",
    "blue": "var(--ui-colors-blue)",
    "red": "var(--ui-colors-red)",
    "text": "var(--ui-colors-text)",
  },
}
`;

exports[`styles() passes variables to style callbacks 2`] = `"-ui-dark-theme"`;

exports[`styles() passes variables to style callbacks 3`] = `
Object {
  "colors": Object {
    "bg": "var(--ui-colors-bg)",
    "blue": "var(--ui-colors-blue)",
    "red": "var(--ui-colors-red)",
    "text": "var(--ui-colors-text)",
  },
}
`;

exports[`styles() passes variables to style callbacks 4`] = `"-ui-light-theme"`;

exports[`styles() passes variables to style callbacks 5`] = `
Object {
  "colors": Object {
    "bg": "var(--ui-colors-bg)",
    "blue": "var(--ui-colors-blue)",
    "lightSpecific": "var(--ui-colors-light-specific)",
    "red": "var(--ui-colors-red)",
    "text": "var(--ui-colors-text)",
  },
}
`;

exports[`styles() replaces disallowed characters in dev labels: 400x400 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-49mne8-box-big{width:400px;height:400px;}
</style>
`;

exports[`styles() returns css styles 1`] = `"display:flex;"`;

exports[`styles() returns css styles 2`] = `"display:flex;display:block;display:inline;"`;

exports[`styles() returns css styles 3`] = `"display:flex;display:inline;"`;

exports[`styles() returns single class name 1`] = `"-ui-1ut9bc3"`;

exports[`styles() returns single class name 2`] = `"-ui-qaw0al"`;

exports[`styles() returns single class name 3`] = `"-ui-broszf"`;

exports[`styles.create() adds nonce to style tags: DOM 1`] = `
<style
  data-dash="-ui"
  nonce="EDNnf03nceIOfn39fn3e9h3sdfa"
>
  
  .-ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
</style>
`;

exports[`styles.create() changes container to document.body 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
</style>
`;

exports[`styles.create() changes key to "css": DOM 1`] = `
<style
  data-dash="css"
>
  
  .css-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
</style>
`;

exports[`styles.create() configures hash algorithm 1`] = `"-ui-e038ab58"`;

exports[`styles.create() turns off vendor prefixing: DOM 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-1ut9bc3{display:flex;}
</style>
`;

exports[`styles.create() turns on speedy 1`] = `
Object {
  ".-ui-1abu38z": Object {
    "0": "display",
    "display": "block",
    "length": 1,
  },
  ".-ui-1ut9bc3": Object {
    "0": "display",
    "display": "flex",
    "length": 1,
  },
}
`;

exports[`styles.global() allows @font-face 1`] = `
<style
  data-dash="-ui"
>
  
  @font-face{font-family:"Open Sans";src:url("/fonts/OpenSans-Regular-webfont.woff2") format("woff2"),url("/fonts/OpenSans-Regular-webfont.woff") format("woff");}
</style>
`;

exports[`styles.global() allows @import 1`] = `
<style
  data-dash="-ui"
>
  
  @import url("navigation.css");
</style>
`;

exports[`styles.global() injects global styles 1`] = `
<style
  data-dash="-ui"
>
  
  html{font-size:100%;}
</style>
`;

exports[`styles.global() passes variables to global styles 1`] = `
Object {
  "colors": Object {
    "bg": "var(--ui-colors-bg)",
    "blue": "var(--ui-colors-blue)",
    "red": "var(--ui-colors-red)",
    "text": "var(--ui-colors-text)",
  },
}
`;

exports[`styles.global() should inject global styles once 1`] = `
<style
  data-dash="-ui"
>
  
  :root{--spacing-0:0;}
</style>
`;

exports[`styles.global() should inject global styles once 2`] = `
<style
  data-dash="-ui"
>
  
  html{font-size:100%;}
</style>
`;

exports[`styles.global() should inject global styles once 3`] = `
<style
  data-dash="-ui"
>
  
  :root{--spacing-1:0.5rem;}
</style>
`;

exports[`styles.global() unmounts global styles w/ callback 1`] = `
<style
  data-dash="-ui"
>
  
  html{font-size:100%;}
</style>
`;

exports[`styles.one() can be called as a function w/ function value 1`] = `"color:var(--ui-color-blue);"`;

exports[`styles.one() can be called as a function w/ string value 1`] = `"-ui-1ut9bc3"`;

exports[`styles.one() creates style w/ template literal 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
</style>
`;

exports[`styles.one() returns a class name when toString() is called 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
</style>
`;

exports[`styles.one() returns css when css() is called 1`] = `"display:flex;"`;

exports[`styles.one() returns css when css.toString() is called 1`] = `"display:flex;"`;

exports[`styles.themes() creates variables: :root 1`] = `
<style
  data-dash="-ui"
>
  
  .-ui-dark-theme{--ui-colors-bg:#000;--ui-colors-text:#fff;}
</style>
`;

exports[`styles.themes() removes variables when eject is called 1`] = `
NodeList [
  <style
    data-dash="-ui"
  >
    
    .-ui-dark-theme{--ui-colors-bg:#000;--ui-colors-text:#fff;}
  </style>,
]
`;

exports[`styles.variables() creates variables w/ scales: :root 1`] = `
<style
  data-dash="-ui"
>
  
  :root{--ui-spacing-0:1rem;--ui-spacing-1:2rem;--ui-spacing-2:4rem;}
</style>
`;

exports[`styles.variables() creates variables: :root 1`] = `
<style
  data-dash="-ui"
>
  
  :root{--ui-colors-blue:#09a;--ui-colors-red:#c12;--ui-colors-light-red:#c1a;--ui-spacing-xs:1rem;--ui-system-p-md:1rem;--ui-system-p-xs:0.25rem;--ui-system-p-sm:0.5rem;--ui-system-p-lg:2rem;--ui-system-p-xl:4rem;}
</style>
`;

exports[`styles.variables() removes variables when eject is called 1`] = `
NodeList [
  <style
    data-dash="-ui"
  >
    
    :root{--ui-colors-blue:#09a;--ui-colors-red:#c12;--ui-colors-light-red:#c1a;--ui-spacing-xs:1rem;--ui-system-p-md:1rem;--ui-system-p-xs:0.25rem;--ui-system-p-sm:0.5rem;--ui-system-p-lg:2rem;--ui-system-p-xl:4rem;}
  </style>,
]
`;
